File: app/ios/MedNote/Services/VisitPipeline.swift
import Foundation

@MainActor
final class VisitPipeline: ObservableObject {
    @Published var step: ProcessingStep?
    @Published var errorMessage: String?

    func processRecording(
        audioURL: URL,
        durationSeconds: Int,
        visitStore: VisitStore,
        settings: SettingsStore,
        doctorLabel: String = "Phone Visit"
    ) async -> Visit? {
        errorMessage = nil
        step = .saving

        var visit = Visit(
            doctor: doctorLabel,
            duration: Visit.formatDuration(seconds: durationSeconds),
            originalLang: settings.recordLanguage.rawValue,
            isProcessing: true
        )

        do {
            let storedName = try visitStore.storeAudio(from: audioURL, visitId: visit.id)
            visit.audioFileName = storedName
        } catch {
            errorMessage = error.localizedDescription
            step = nil
            return nil
        }

        visitStore.add(visit)
        step = .transcribing

        do {
            let result = try await AIService.runPipeline(
                audioURL: visitStore.audioURL(for: visit)!,
                groqKey: settings.groqKey,
                geminiKey: settings.geminiKey,
                userLanguage: settings.language,
                recordLanguage: settings.recordLanguage,
                durationSeconds: durationSeconds
            )

            step = .analyzing
            visit.originalTranscript = result.transcriptLines
            visit.summary = result.summary
            visit.translatedTranscript = result.translatedTranscript
            visit.translatedSummary = settings.language != .en ? result.summary : nil
            visit.extractedItems = result.extractedItems
            visit.isProcessing = false
            visitStore.update(visit)
            step = .done
            try? await Task.sleep(nanoseconds: 600_000_000)
            step = nil
            return visit
        } catch {
            errorMessage = error.localizedDescription
            visit.isProcessing = false
            visit.summary = "Processing failed: \(error.localizedDescription)"
            visitStore.update(visit)
            step = nil
            return visit
        }
    }

    func processImportedFile(
        url: URL,
        visitStore: VisitStore,
        settings: SettingsStore
    ) async -> Visit? {
        await processRecording(
            audioURL: url,
            durationSeconds: 0,
            visitStore: visitStore,
            settings: settings,
            doctorLabel: "Imported Recording"
        )
    }
}

File: app/android/app/src/main/kotlin/com/mednote/service/VisitPipeline.kt
package com.mednote.service

import com.mednote.data.SettingsStore
import com.mednote.data.VisitStore
import com.mednote.models.AppLanguage
import com.mednote.models.ProcessingStep
import com.mednote.models.Visit
import com.mednote.models.VisitFormats
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.withContext
import java.io.File

class VisitPipeline {
    private val _step = MutableStateFlow<ProcessingStep?>(null)
    val step: StateFlow<ProcessingStep?> = _step.asStateFlow()

    private val _error = MutableStateFlow<String?>(null)
    val error: StateFlow<String?> = _error.asStateFlow()

    suspend fun processRecording(
        audioFile: File,
        durationSeconds: Int,
        visitStore: VisitStore,
        settings: SettingsStore,
        doctorLabel: String = "Phone Visit"
    ): Visit? {
        _error.value = null
        _step.value = ProcessingStep.SAVING

        var visit = Visit(
            doctor = doctorLabel,
            duration = VisitFormats.formatDuration(durationSeconds),
            originalLang = settings.recordLanguage.first().code,
            isProcessing = true
        )

        val storedName = runCatching {
            visitStore.storeAudio(audioFile, visit.id)
        }.getOrElse {
            _error.value = it.message
            _step.value = null
            return null
        }
        visit = visit.copy(audioFileName = storedName)
        visitStore.add(visit)

        return runAiPipeline(visit, visitStore, settings)
    }

    suspend fun processImportedFile(
        uriFile: File,
        visitStore: VisitStore,
        settings: SettingsStore
    ): Visit? = processRecording(
        audioFile = uriFile,
        durationSeconds = 0,
        visitStore = visitStore,
        settings = settings,
        doctorLabel = "Imported Recording"
    )

    private suspend fun runAiPipeline(
        initialVisit: Visit,
        visitStore: VisitStore,
        settings: SettingsStore
    ): Visit? {
        _step.value = ProcessingStep.TRANSCRIBING
        val audio = visitStore.audioFile(initialVisit) ?: return fail(initialVisit, visitStore, "Audio file missing")

        return try {
            val groq = settings.groqKey.first()
            val gemini = settings.geminiKey.first()
            val lang = settings.language.first()
            val recLang = settings.recordLanguage.first()

            val result = withContext(Dispatchers.IO) {
                AIService.runPipeline(audio, groq, gemini, lang, recLang)
            }

            _step.value = ProcessingStep.ANALYZING
            val updated = initialVisit.copy(
                originalTranscript = result.transcriptLines,
                summary = result.summary,
                translatedTranscript = result.translatedTranscript,
                translatedSummary = if (lang != AppLanguage.EN) result.summary else null,
                extractedItems = result.extractedItems,
                isProcessing = false
            )
            visitStore.update(updated)
            _step.value = ProcessingStep.DONE
            kotlinx.coroutines.delay(600)
            _step.value = null
            updated
        } catch (e: Exception) {
            fail(initialVisit, visitStore, e.message ?: "Processing failed")
        }
    }

    private fun fail(visit: Visit, visitStore: VisitStore, message: String): Visit {
        _error.value = message
        val updated = visit.copy(
            isProcessing = false,
            summary = "Processing failed: $message"
        )
        visitStore.update(updated)
        _step.value = null
        return updated
    }
}
